server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -1146,9 +1160,9 @@ private void initializeExecutors()
     if (persistExecutor == null) {
       // use a blocking single threaded executor to throttle the firehose when write to disk is slow
       persistExecutor = MoreExecutors.listeningDecorator(
-          Execs.newBlockingSingleThreaded(
+          Execs.newBlockingThreaded(
               "[" + StringUtils.encodeForFormat(myId) + "]-appenderator-persist",
-              maxPendingPersists
+              persistThreads, maxPendingPersists
```

Comment:
I'm wary of the `persistExecutor` being called by callers assuming it is single-threaded. For example, there's nothing I can see that prevents the `close()` from occurring while a task is still running. Nothing stands out as obviously broken though. But, I do see comments like this:

```
        // use persistExecutor to make sure that all the pending persists completes before
        // starting to abandon segments
        persistExecutor
```
Comments suggest that - if we have multiple threads - we might start processing an abandon request before or while a persist request is in queue. What scenarios/mitigations are there here?

Reply:
I see will get back to you on it

Reply:
Was looking at this, wouldn't the shutdown on the executor interrupt any running threads anyway. I mean even with single thread the pending persist would not complete. Also when abandoning the segments are not going to get pushed to deep storage and nor would the kafka offsets be committed isn't it. So any pending persists in the queue wouldn't matter.